doc/cprover-manual/contracts-decreases.md

```diff
@@ -1,3 +1,183 @@
-# Decreases Clause
+# Decreases Clauses
 
-TODO: Document `__CPROVER_decreases`
+A _decreases_ clause specifies a measure that must strictly decrease
+monotonically at every iteration of a loop.
+By demonstrating that the measure
+1. is bounded from below, and
+2. strictly decreases monotonically at each iteration,
+we can prove termination of loops.
+This is because the measure must eventually hit the lower bound
+at which point the loop must terminate,
+since the measure cannot strictly decrease further.
+
+In the scientific literature, this measure is also called a _loop variant_ or _ranking function_.
+Although they can be arbitrary (strict) well-founded relations on arbitrary domains,
+we restrict decreases clauses to (strict) arithmetic inequalities.
+Unlike mathematical integers, arithmetic types in C are all bounded from below.
+Furthermore, unlike mathematical real numbers, floating-point arithmetic types in C are not dense.
+Hence, any strictly decreasing arithmetic expression in C can be used as a decreases clause.
+
+### Syntax
+
+A one-dimensional (1D) decreases clause for a loop is an arithmetic expression `e`
+over the variables visible at the same scope as the loop,
+specified as `__CPROVER_decreases(e)`.
+
+Like invariant clauses, decreases clauses may be specified just after the loop guard.
+An example of a 1D decreases clause is shown below.
+
+```c
+for(int i = 0; i < n; i += 2)
+__CPROVER_loop_invariant(0 <= i && i <= n)
+__CPROVER_decreases(n - i)
+{ ... }
+```
+
+Please see the [invariant clauses](contracts-invariants.md) page
+for more examples on `for` and `do...while` loops.
+
+To help prove termination of more complex loops,
+CBMC also supports multi-dimensional decreases clauses.
+A multi-dimensional decreases clause is an [ordered tuple](https://en.wikipedia.org/wiki/Tuple)
+of arithmetic expressions, specified as `__CPROVER_decreases(e_1, e_2, ..., e_n)`.
+An example of a multi-dimensional decreases clause is given below.
+
+```c
+while(i < n)
+__CPROVER_loop_invariant(0 <= i && i <= n)
+__CPROVER_loop_invariant(0 <= j && j <= n)
+__CPROVER_decreases(n - i, n - j)
+{
+  if (j < n)
+    j++;
+  else
+  {
+    i++;
+    j = 0;
+  }
+}
+```
+
+We extend the strict arithmetic comparison for 1D decreases clauses
+to a strict [lexicographic comparison](https://en.wikipedia.org/wiki/Lexicographic_order)
+for multi-dimensional decreases clauses.
+
+**Important.**
+Like invariant clauses, decreases clauses must be free of side effects,
+for example, mutation of local or global variables.
```

Comment:
Do we check that? Are we telling people to take care here lest they produce an unsound result, or are we telling them that if they make this mistake the tool will complain with an error. If so, it might be nice to describe the error they will get.

Reply:
Yes, we do check this and generate an error as shown in this regression test:
https://github.com/diffblue/cbmc/blob/3e0007f9ed818969185676eb96f8811dc75e8705/regression/contracts/variant_side_effects_fail/test.desc#L4

I would add this message to this doc & the invariants doc.